test/object-store/sync/flx_sync.cpp

```diff
@@ -252,25 +266,42 @@ TEST_CASE("flx: uploading an object that is out-of-view results in a client rese
 
             CppContext c(realm);
             realm->begin_transaction();
+            auto valid_obj = ObjectId::gen();
+            auto invalid_obj = ObjectId::gen();
             Object::create(c, realm, "TopLevel",
-                           util::Any(AnyDict{{"_id", ObjectId::gen()},
+                           util::Any(AnyDict{{"_id", valid_obj},
                                              {"queryable_str_field", std::string{"foo"}},
                                              {"queryable_int_field", static_cast<int64_t>(5)},
                                              {"non_queryable_field", std::string{"non queryable 1"}}}));
             Object::create(c, realm, "TopLevel",
-                           util::Any(AnyDict{{"_id", ObjectId::gen()},
+                           util::Any(AnyDict{{"_id", invalid_obj},
                                              {"queryable_str_field", std::string{"bar"}},
                                              {"queryable_int_field", static_cast<int64_t>(10)},
                                              {"non_queryable_field", std::string{"non queryable 2"}}}));
             realm->commit_transaction();
 
             auto sync_error = std::move(error_future).get();
-            CHECK(sync_error.error_code == sync::make_error_code(sync::ProtocolError::write_not_allowed));
+            CHECK(sync_error.error_code == sync::make_error_code(sync::ProtocolError::compensating_write));
             CHECK(sync_error.is_session_level_protocol_error());
-            CHECK(sync_error.is_client_reset_requested());
+            CHECK(!sync_error.is_client_reset_requested());
+            CHECK(sync_error.compensating_writes_info.size() == 1);
+            auto write_info = sync_error.compensating_writes_info[0];
+            CHECK(write_info.primary_key.is_type(type_ObjectId));
+            CHECK(write_info.primary_key.get_object_id() == invalid_obj);
+            CHECK(write_info.object_name == "TopLevel");
+            CHECK(write_info.reason.find("object is outside of the current query view") != std::string::npos);
+
+
+            wait_for_download(*realm);
+            realm->refresh();
+
+            auto top_level_table = realm->read_group().get_table("class_TopLevel");
+            REQUIRE(top_level_table->size() == 1);
+            REQUIRE(top_level_table->get_object_with_primary_key(valid_obj));
         });
     }
 }
+#endif
```

Comment:
I think another interesting test would be to make a modification to an embedded object for which the server disallows modifications to the root object, and ensure that the root pk gets reported to the client.

Reply:
i added this test